bee-crypto/src/ternary/bigint/u384/mod.rs

```diff
@@ -0,0 +1,482 @@
+where
+    E: fmt::Debug,
+    R: BinaryRepresentation<Inner = D>,
+    D: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("U384")
+            .field("inner", &self.inner.iter())
+            .field("_phantom", &self._phantom)
+            .finish()
+    }
+}
+
+impl U384<BigEndian, U32Repr> {
+    /// Reinterprets the U384 as an I384.
+    pub fn as_i384(self) -> I384<BigEndian, U32Repr> {
+        I384::<BigEndian, U32Repr>::from_array(self.inner)
+    }
+
+    /// Shifts the U384 in signed space.
+    pub fn shift_into_i384(mut self) -> I384<BigEndian, U32Repr> {
+        self.sub_inplace(*BE_U32_HALF_MAX);
+        self.sub_inplace(Self::one());
+        self.as_i384()
+    }
+
+    /// Adds `other` onto `self` in place.
+    pub fn add_inplace(&mut self, other: Self) {
+        let mut overflown = false;
+        let self_iter = self.inner.iter_mut().rev();
+        let other_iter = other.inner.iter().rev();
+
+        for (s, o) in self_iter.zip(other_iter) {
+            let (sum, still_overflown) = s.overflowing_add_with_carry(*o, overflown as u32);
+            *s = sum;
+            overflown = still_overflown;
+        }
+    }
+
+    /// Adds `other` in place, returning the number of digits required accomodate `other` (starting from the least
+    /// significant one).
+    pub fn add_digit_inplace<T: Into<u32>>(&mut self, other: T) -> usize {
+        let other = other.into();
+
+        let mut i = self.inner.len() - 1;
+
+        let (sum, mut overflown) = self.inner[i].overflowing_add(other);
+        self.inner[i] = sum;
+
+        i -= 1;
+
+        while overflown {
+            let (sum, still_overflown) = self.inner[i].overflowing_add(1u32);
+            self.inner[i] = sum;
+            overflown = still_overflown;
+            i -= 1;
+        }
+
+        i
+    }
+
+    /// Divides the u384 by 2 by bitshifting all bits one position to the right.
+    pub fn divide_by_two(&mut self) {
+        let mut i = self.inner.len() - 1;
+        while i < self.inner.len() - 1 {
+            let (left_slice, right_slice) = self.inner.split_at_mut(i + 1);
+            let left = &mut left_slice[i];
+            let right = &mut right_slice[0];
+            *left >>= 1;
+            *left |= *right << 31;
+            i -= 1;
+        }
+        self.inner[0] >>= 1;
+    }
+
+    /// Creates an U384 from an unbalanced T242.
+    pub fn from_t242(trits: T242<Utrit>) -> Self {
+        let u384_le = U384::<LittleEndian, U32Repr>::from_t242(trits);
+        u384_le.into()
+    }
+
+    /// Subtract `other` from `self` inplace.
+    ///
+    /// This function is defined in terms of `overflowing_add` by making use of the following identity (in terms of
+    /// Two's complement, and where `!` is logical bitwise negation):
+    ///
+    /// !x = -x - 1 => -x = !x + 1
+    pub fn sub_inplace(&mut self, other: Self) {
+        let self_iter = self.inner.iter_mut().rev();
```

Comment:
I'm not sure why these iterators are borrowed like this... why not just use them directly in the iterator?

Reply:
Changed